executor/write.go

```diff
@@ -65,50 +70,53 @@ func updateRecord(ctx context.Context, h int64, oldData, newData []types.Datum,
 			}
 			t.RebaseAutoID(val, true)
 		}
-		casted, err := table.CastValue(ctx, newData[i], col.ToInfo())
+		cmp, err := newData[i].CompareDatum(sc, oldData[i])
 		if err != nil {
 			return false, errors.Trace(err)
 		}
-		newData[i] = casted
-		touched[i] = true
-		assignExists = true
+		if cmp != 0 {
+			changed = true
+			touched[i] = true
+			if col.IsPKHandleColumn(t.Meta()) {
+				handleChanged = true
+			}
+		} else {
+			if mysql.HasOnUpdateNowFlag(col.Flag) && touched[i] {
+				// It's for "UPDATE t SET ts = ts" and ts is a timestamp.
+				onUpdateNoChange[i] = true
+			}
+			touched[i] = false
+		}
 	}
-
-	// If no assign list for this table, no need to update.
-	if !assignExists {
-		return false, nil
+	if changed {
+		// Only if some fields are really changed, we can change those on-update fields.
+		for i, col := range t.Cols() {
+			if mysql.HasOnUpdateNowFlag(col.Flag) && !touched[i] && !onUpdateNoChange[i] {
+				v, err := expression.GetTimeValue(ctx, expression.CurrentTimestamp, col.Tp, col.Decimal)
+				if err != nil {
+					return false, errors.Trace(err)
+				}
+				newData[i] = v
+			} else {
+				continue
+			}
+		}
 	}
```

Comment:
Move line 92-103 behind line 118.  Then you can move the judgment condition of `changed`.